api/v1beta1/content.go

```diff
@@ -0,0 +1,127 @@
+package v1beta1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GrafanaResourceDatasource is used to set the datasource name of any templated datasources in
+// content definitions (e.g., dashboard JSON).
+type GrafanaContentDatasource struct {
+	InputName      string `json:"inputName"`
+	DatasourceName string `json:"datasourceName"`
+}
+
+type GrafanaContentEnv struct {
+	Name string `json:"name"`
+	// Inline env value
+	// +optional
+	Value string `json:"value,omitempty"`
+	// Reference on value source, might be the reference on a secret or config map
+	// +optional
+	ValueFrom GrafanaContentEnvFromSource `json:"valueFrom,omitempty"`
+}
+
+type GrafanaContentEnvFromSource struct {
+	// Selects a key of a ConfigMap.
+	// +optional
+	ConfigMapKeyRef *v1.ConfigMapKeySelector `json:"configMapKeyRef,omitempty"`
+	// Selects a key of a Secret.
+	// +optional
+	SecretKeyRef *v1.SecretKeySelector `json:"secretKeyRef,omitempty"`
+}
+
+type GrafanaContentUrlBasicAuth struct {
+	Username *v1.SecretKeySelector `json:"username,omitempty"`
+	Password *v1.SecretKeySelector `json:"password,omitempty"`
+}
+
+type GrafanaContentUrlAuthorization struct {
+	BasicAuth *GrafanaContentUrlBasicAuth `json:"basicAuth,omitempty"`
+}
+
+type JsonnetProjectBuild struct {
+	JPath              []string `json:"jPath,omitempty"`
+	FileName           string   `json:"fileName"`
+	GzipJsonnetProject []byte   `json:"gzipJsonnetProject"`
+}
+
+// GrafanaComContentReference is a reference to content hosted on grafana.com
+type GrafanaComContentReference struct {
+	Id       int  `json:"id"`
+	Revision *int `json:"revision,omitempty"`
+}
+
+type GrafanaContentSpec struct {
+	// Manually specify the uid, overwrites uids already present in the json model.
+	// Can be any string consisting of alphanumeric characters, - and _ with a maximum length of 40.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec.uid is immutable"
+	// +kubebuilder:validation:MaxLength=40
+	// +kubebuilder:validation:Pattern="^[a-zA-Z0-9-_]+$"
+	CustomUID string `json:"uid,omitempty"`
```

Comment:
I feel like this should stay in the `GrafanaDashboardSpec` as it doesn't affect the way the model is fetched

Reply:
yeah, it was an interesting challenge to figure out which parts to lift out. the reason i ultimately decided to place CustomUID here is that library elements (and i think most other Grafana entities) have this UID vs ID concept. library panels in particular are referenced by UID, so i thought this feature of dashboards was useful and translated well to library panels.

i wound up moving the IsUpdatedUID and CustomUIDOrID functions to the [resolver](https://github.com/grafana/grafana-operator/pull/1845/commits/5867bef612b0115e0d72aae9a2170ecdd97021b5#diff-5723394742dbfbf05333333388ae1b0de1f9f47650f514bfba6b5d91ace771aaR19-R49) to make them generic as a result.